Constructor injection in JAX-RS Resource using OpenApi Annotations causes RuntimeException

**Describe the bug**
JAX-RS resources which have constructor injection and contains openapi annotations causes RuntimeException, this issued is related to #10430

**Expected behavior**
Constructor injection also work like the field injection.

**Actual behavior**
On startup:
Caused by: java.lang.RuntimeException: RESTEASY003190: Could not find constructor for class: xx.xxx.IncidentResource

**To Reproduce**
Is it supposed to be fixed on 1.6.1.Final? I'm using this version and still having the issue, [here a sample](https://github.com/mageddo/poc/tree/1128295/quarkus-swagger-issue)

Explicitly add a NonArg constructor fixes the problem, I expect Quarkus to solve it for me though

```java
@Tag(name = "hello-samples")
@Path("/hello")
public class FruitResource {

    HelloService helloService;

    public FruitResource() {
    }

    @Inject
    public FruitResource(HelloService helloService) {
        this.helloService = helloService;
    }
....
```

--- a/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
+++ b/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
@@ -75,6 +75,7 @@ import io.quarkus.resteasy.server.common.runtime.QuarkusResteasyDeployment;
 import io.quarkus.resteasy.server.common.spi.AdditionalJaxRsResourceDefiningAnnotationBuildItem;
 import io.quarkus.resteasy.server.common.spi.AdditionalJaxRsResourceMethodAnnotationsBuildItem;
 import io.quarkus.resteasy.server.common.spi.AdditionalJaxRsResourceMethodParamAnnotations;
+import io.quarkus.resteasy.server.common.spi.AllowedJaxRsAnnotationPrefixBuildItem;
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigRoot;
 import io.quarkus.undertow.deployment.FilterBuildItem;
@@ -174,6 +175,7 @@ public class ResteasyServerCommonProcessor {
             List<AdditionalJaxRsResourceDefiningAnnotationBuildItem> additionalJaxRsResourceDefiningAnnotations,
             List<AdditionalJaxRsResourceMethodAnnotationsBuildItem> additionalJaxRsResourceMethodAnnotations,
             List<AdditionalJaxRsResourceMethodParamAnnotations> additionalJaxRsResourceMethodParamAnnotations,
+            List<AllowedJaxRsAnnotationPrefixBuildItem> friendlyJaxRsAnnotationPrefixes,
             List<ResteasyDeploymentCustomizerBuildItem> deploymentCustomizers,
             JaxrsProvidersToRegisterBuildItem jaxrsProvidersToRegisterBuildItem,
             CombinedIndexBuildItem combinedIndexBuildItem,
@@ -283,7 +285,8 @@ public class ResteasyServerCommonProcessor {
 
         // generate default constructors for suitable concrete @Path classes that don't have them
         // see https://issues.jboss.org/browse/RESTEASY-2183
-        generateDefaultConstructors(transformers, withoutDefaultCtor, additionalJaxRsResourceDefiningAnnotations);
+        generateDefaultConstructors(transformers, withoutDefaultCtor, additionalJaxRsResourceDefiningAnnotations,
+                friendlyJaxRsAnnotationPrefixes);
 
         checkParameterNames(beanArchiveIndexBuildItem.getIndex(), additionalJaxRsResourceMethodParamAnnotations);
 
@@ -385,6 +388,18 @@ public class ResteasyServerCommonProcessor {
             }
         }));
         resteasyDeployment.produce(new ResteasyDeploymentBuildItem(path, deployment));
+    }
+
+    @BuildStep
+    List<AllowedJaxRsAnnotationPrefixBuildItem> registerCompatibleAnnotationPrefixes() {
+        List<AllowedJaxRsAnnotationPrefixBuildItem> prefixes = new ArrayList<>();
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem(packageName(ResteasyDotNames.PATH)));
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem("kotlin")); // make sure the annotation that the Kotlin compiler adds don't interfere with creating a default constructor
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem("lombok")); // same for lombok
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem("io.quarkus.security")); // same for the security annotations
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem("javax.annotation.security"));
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem("jakarta.annotation.security"));
+        return prefixes;
     }
 
     @BuildStep
@@ -579,15 +594,14 @@ public class ResteasyServerCommonProcessor {
 
     private static void generateDefaultConstructors(BuildProducer<BytecodeTransformerBuildItem> transformers,
             Map<DotName, ClassInfo> withoutDefaultCtor,
-            List<AdditionalJaxRsResourceDefiningAnnotationBuildItem> additionalJaxRsResourceDefiningAnnotations) {
+            List<AdditionalJaxRsResourceDefiningAnnotationBuildItem> additionalJaxRsResourceDefiningAnnotations,
+            List<AllowedJaxRsAnnotationPrefixBuildItem> friendlyJaxRsAnnotationPrefixes) {
 
         final Set<String> allowedAnnotationPrefixes = new HashSet<>(1 + additionalJaxRsResourceDefiningAnnotations.size());
-        allowedAnnotationPrefixes.add(packageName(ResteasyDotNames.PATH));
-        allowedAnnotationPrefixes.add("kotlin"); // make sure the annotation that the Kotlin compiler adds don't interfere with creating a default constructor
-        allowedAnnotationPrefixes.add("lombok"); // same for lombok
-        allowedAnnotationPrefixes.add("io.quarkus.security"); // same for the security annotations
-        allowedAnnotationPrefixes.add("javax.annotation.security");
-        allowedAnnotationPrefixes.add("jakarta.annotation.security");
+        friendlyJaxRsAnnotationPrefixes.stream()
+                .map(prefix -> prefix.getAnnotationPrefix())
+                .forEachOrdered(allowedAnnotationPrefixes::add);
+
         for (AdditionalJaxRsResourceDefiningAnnotationBuildItem additionalJaxRsResourceDefiningAnnotation : additionalJaxRsResourceDefiningAnnotations) {
             final String packageName = packageName(additionalJaxRsResourceDefiningAnnotation.getAnnotationClass());
             if (packageName != null) {
@@ -605,7 +619,7 @@ public class ResteasyServerCommonProcessor {
             boolean hasNonJaxRSAnnotations = false;
             for (AnnotationInstance instance : classInfo.classAnnotations()) {
                 final String packageName = packageName(instance.name());
-                if (packageName == null || !allowedAnnotationPrefixes.contains(packageName)) {
+                if (packageName == null || !isPackageAllowed(allowedAnnotationPrefixes, packageName)) {
                     hasNonJaxRSAnnotations = true;
                     break;
                 }
@@ -641,6 +655,10 @@ public class ResteasyServerCommonProcessor {
                         }
                     }));
         }
+    }
+
+    private static boolean isPackageAllowed(Set<String> allowedAnnotationPrefixes, String packageName) {
+        return allowedAnnotationPrefixes.stream().anyMatch(prefix -> packageName.startsWith(prefix));
     }
 
     private static String packageName(DotName dotName) {

--- a/extensions/resteasy-server-common/spi/src/main/java/io/quarkus/resteasy/server/common/spi/AllowedJaxRsAnnotationPrefixBuildItem.java
+++ b/extensions/resteasy-server-common/spi/src/main/java/io/quarkus/resteasy/server/common/spi/AllowedJaxRsAnnotationPrefixBuildItem.java
@@ -1,0 +1,20 @@
+package io.quarkus.resteasy.server.common.spi;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * The package prefix of any annotations that have to be compatible with JaxRs resource class
+ * to allow constructor injection.
+ */
+public final class AllowedJaxRsAnnotationPrefixBuildItem extends MultiBuildItem {
+
+    private final String prefix;
+
+    public AllowedJaxRsAnnotationPrefixBuildItem(String prefix) {
+        this.prefix = prefix;
+    }
+
+    public String getAnnotationPrefix() {
+        return prefix;
+    }
+}

--- a/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
+++ b/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
@@ -51,6 +51,7 @@ import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.quarkus.deployment.logging.LogCleanupFilterBuildItem;
+import io.quarkus.resteasy.server.common.spi.AllowedJaxRsAnnotationPrefixBuildItem;
 import io.quarkus.resteasy.server.common.spi.ResteasyJaxrsConfigBuildItem;
 import io.quarkus.runtime.LaunchMode;
 import io.quarkus.smallrye.openapi.common.deployment.SmallRyeOpenApiConfig;
@@ -157,6 +158,13 @@ public class SmallRyeOpenApiProcessor {
                 new FilteredIndexView(
                         compositeIndex,
                         new OpenApiConfigImpl(ConfigProvider.getConfig())));
+    }
+
+    @BuildStep
+    public List<AllowedJaxRsAnnotationPrefixBuildItem> registerJaxRsSupportedAnnotation() {
+        List<AllowedJaxRsAnnotationPrefixBuildItem> prefixes = new ArrayList<>();
+        prefixes.add(new AllowedJaxRsAnnotationPrefixBuildItem("org.eclipse.microprofile.openapi.annotations"));
+        return prefixes;
     }
 
     @BuildStep

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiDefaultPathTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiDefaultPathTestCase.java
@@ -15,7 +15,7 @@ public class OpenApiDefaultPathTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class));
+                    .addClasses(OpenApiResource.class, ResourceBean.class));
 
     @Test
     public void testOpenApiPathAccessResource() {
@@ -34,7 +34,11 @@ public class OpenApiDefaultPathTestCase {
                 .header("Content-Type", "application/json;charset=UTF-8")
                 .body("openapi", Matchers.startsWith("3.0"))
                 .body("info.title", Matchers.equalTo("Generated API"))
-                .body("paths", Matchers.hasKey("/resource"));
+                .body("tags.name[0]", Matchers.equalTo("test"))
+                .body("paths.'/resource'.get.servers[0]", Matchers.hasKey("url"))
+                .body("paths.'/resource'.get.security[0]", Matchers.hasKey("securityRequirement"))
+                .body("paths.'/resource'.get", Matchers.hasKey("openApiExtension"));
+
         RestAssured.given()
                 .when().options(OPEN_API_PATH)
                 .then().header("Access-Control-Allow-Methods", "GET, HEAD, OPTIONS");

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiHttpRootDefaultPathTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiHttpRootDefaultPathTestCase.java
@@ -16,7 +16,7 @@ public class OpenApiHttpRootDefaultPathTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsResource(new StringAsset("quarkus.http.root-path=/foo"), "application.properties"));
 
     @Test

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiPathWithSegmentsTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiPathWithSegmentsTestCase.java
@@ -15,7 +15,7 @@ public class OpenApiPathWithSegmentsTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsResource(new StringAsset("quarkus.smallrye-openapi.path=" + OPEN_API_PATH),
                             "application.properties"));
 

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiPathWithoutSegmentsTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiPathWithoutSegmentsTestCase.java
@@ -15,7 +15,7 @@ public class OpenApiPathWithoutSegmentsTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsResource(new StringAsset("quarkus.smallrye-openapi.path=" + OPEN_API_PATH),
                             "application.properties"));
 

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiResource.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiResource.java
@@ -1,15 +1,32 @@
 package io.quarkus.smallrye.openapi.test.jaxrs;
 
+import javax.inject.Inject;
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 import javax.ws.rs.QueryParam;
 
+import org.eclipse.microprofile.openapi.annotations.extensions.Extension;
+import org.eclipse.microprofile.openapi.annotations.security.SecurityRequirement;
+import org.eclipse.microprofile.openapi.annotations.servers.Server;
+import org.eclipse.microprofile.openapi.annotations.tags.Tag;
+
 @Path("/resource")
+@Tag(name = "test")
+@Extension(name = "openApiExtension", value = "openApiExtensionValue")
+@SecurityRequirement(name = "securityRequirement", scopes = "securityRequirementScope")
+@Server(url = "serverUrl")
 public class OpenApiResource {
+
+    private ResourceBean resourceBean;
+
+    @Inject
+    public OpenApiResource(ResourceBean resourceBean) {
+        this.resourceBean = resourceBean;
+    }
 
     @GET
     public String root() {
-        return "resource";
+        return resourceBean.toString();
     }
 
     @GET

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiWithConfigTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiWithConfigTestCase.java
@@ -16,7 +16,7 @@ public class OpenApiWithConfigTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsManifestResource("test-openapi.yaml", "openapi.yaml")
                     .addAsResource(new StringAsset("mp.openapi.scan.disable=true\nmp.openapi.servers=https://api.acme.org/"),
                             "application.properties"));

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiWithResteasyPathHttpRootDefaultPathTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiWithResteasyPathHttpRootDefaultPathTestCase.java
@@ -16,7 +16,7 @@ public class OpenApiWithResteasyPathHttpRootDefaultPathTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsResource(new StringAsset("quarkus.http.root-path=/http-root-path\n" +
                             "quarkus.resteasy.path=/resteasy-path"),
                             "application.properties"));

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiWithResteasyPathTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/OpenApiWithResteasyPathTestCase.java
@@ -16,7 +16,7 @@ public class OpenApiWithResteasyPathTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsResource(new StringAsset("quarkus.resteasy.path=/foo/bar"),
                             "application.properties"));
 

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/ResourceBean.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/ResourceBean.java
@@ -1,0 +1,11 @@
+package io.quarkus.smallrye.openapi.test.jaxrs;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class ResourceBean {
+    @Override
+    public String toString() {
+        return "resource";
+    }
+}

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/SwaggerAndOpenAPIWithCommonPrefixTest.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/SwaggerAndOpenAPIWithCommonPrefixTest.java
@@ -19,7 +19,7 @@ public class SwaggerAndOpenAPIWithCommonPrefixTest {
     @RegisterExtension
     static final QuarkusUnitTest config = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClass(OpenApiResource.class)
+                    .addClasses(OpenApiResource.class, ResourceBean.class)
                     .addAsResource(new StringAsset("quarkus.smallrye-openapi.path=/swagger"), "application.properties"));
 
     @Test

